neqo-transport/src/recovery/sent.rs

```diff
@@ -196,19 +196,60 @@ impl SentPackets {
         R::IntoIter: ExactSizeIterator,
     {
         let mut result = Vec::new();
-        // Remove all packets. We will add them back as we don't need them.
+
+        // Start with all packets. We will add unacknowledged packets back.
+        //  [---------------------------packets----------------------------]
         let mut packets = std::mem::take(&mut self.packets);
+
+        let mut previous_range_start: Option<PacketNumber> = None;
+
         for range in acked_ranges {
-            // For each acked range, split off the acknowledged part,
-            // then split off the part that hasn't been acknowledged.
-            // This order works better when processing ranges that
-            // have already been processed, which is common.
-            let mut acked = packets.split_off(range.start());
-            let keep = acked.split_off(&(*range.end() + 1));
-            self.packets.extend(keep);
-            result.extend(acked.into_values().rev());
+            // Split off at the end of the acked range.
+            //
+            //  [---------packets--------][----------after_acked_range---------]
+            let after_acked_range = packets.split_off(&(*range.end() + 1));
+
+            // Split off at the start of the acked range.
+            //
+            //  [-packets-][-acked_range-][----------after_acked_range---------]
+            let acked_range = packets.split_off(range.start());
+
+            // According to RFC 9000 19.3.1 ACK ranges are in descending order:
+            //
+            // > Each ACK Range consists of alternating Gap and ACK Range Length
+            // > values in **descending packet number order**.
+            //
+            // <https://www.rfc-editor.org/rfc/rfc9000.html#section-19.3.1>
+            if let Some(prev_start) = previous_range_start {
+                debug_assert!(
+                    *range.end() < prev_start,
```

Comment:
Given that ACK ranges are encoded on the wire relative to the previous range they can not overlap.

Given that `range` is inclusive, its length can not be zero.

Thus we need a `<` and not `<=`.

Please correct me if I am missing something.

Reply:
Not saying your analysis is false, but shouldn't we've hit the assert in that case?

Reply:
Would `debug_assert!(prev_start.is_none_or(|s| s > *range.end()))` work better?

I don't think that you need a big print statement: it will never happen, except in a test.